(maint) Use beaker-hostgenerator to create nodesets

justinstoller opened the pull request:
This pulls in beaker-hostgenerator and adds a new rake task
`gen_nodeset` for utilizing it. The output of the rake task on success
is the location of the generated nodeset for use when calling Beaker
later on.

Example workflow:
```
export BEAKER_nodeset=`bundle exec rake gen_nodeset`
bundle exec rake beaker
```

igalic commented:
i'm scared.

ferventcoder commented:
@igalic nothing to be scared of, this helps in acceptance testing with beaker - https://github.com/puppetlabs/beaker-hostgenerator

ferventcoder commented on config_defaults.yml:
```diff
@@ -24,9 +24,14 @@ Gemfile:
       - gem: beaker-puppet_install_helper
       - gem: master_manipulator
+      - gem: beaker-hostgenerator
+        version: '*location_for(ENV["BEAKER_VERSION"])'
```
this stays in sync with the beaker version?

justinstoller replied:
Good catch, updated to use the same env var used in Puppet!

DavidS commented on moduleroot/Gemfile:
```diff
@@ -1,12 +1,14 @@
+def location_for(place, version = nil)
   if place =~ /^(git[:@][^#]*)#(.*)/
-    [fake_version, { :git => $1, :branch => $2, :require => false }].compact
+    [fake_version, { :git => $1, :branch => $2, :require => false}].compact
```
fake_version is not defined, probably also use `version` here?

justinstoller replied:
Good catch!

DavidS commented on moduleroot/Gemfile:
```diff
@@ -20,7 +22,7 @@ group <%= group %> do
 <%            { -%>
 <%              'gem'     => gem['gem'], -%>
-<%              'version' => gem['version'], -%>
+<%              'version' => gem['version'].to_s =~ /(^$)|(location_for)/ ? gem['version'] : "'#{gem['version']}'", -%>
```
I'd almost prefer using a different key in the config here, instead of :version, so that the templatae renders the ENV , instead of passing through random ruby code.

so in the config_defaults:

```
- gem: beaker-rspec
  from_env: BEAKER_RSPEC_VERSION
```

and here in the Gemfile.erb add `location_for(ENV["<%= gem['from_env'] %>"]` instead.

justinstoller replied:
That's a great idea, I've taken it whole cloth in an update

DavidS commented on moduleroot/Rakefile:
```diff
@@ -7,3 +7,27 @@ PuppetLint.configuration.fail_on_warnings = true
+  require 'securerandom'
+  require 'fileutils'
+
+  agent_target = ENV['TEST_TARGET']
```
This should have a (overridable) default. I don't think users want to be forced to set this always

justinstoller replied:
Good idea. I couldn't help but think that if we're setting a default we should warn the user what we're doing. I've got an update incoming that does both.

DavidS commented on moduleroot/Rakefile:
```diff
@@ -7,3 +7,27 @@ PuppetLint.configuration.fail_on_warnings = true
 PuppetLint.configuration.ignore_paths = ["spec/**/*.pp", "pkg/**/*.pp"]
+
+desc 'Generate pooler nodesets'
+task :gen_nodeset do
```
Mulling over this, I wonder if it wouldn't be better suited to hoock this up directly to https://github.com/puppetlabs/puppetlabs_spec_helper/blob/master/lib/puppetlabs_spec_helper/rake_tasks.rb#L14

justinstoller replied:
I think moving it into PSH would be great if beaker-hostgenerator had vagrant support and was generally applicable beyond Puppet's infra. But it doesn't so I don't think, at this time, PSH is the best place for it. However, when vagrant support is added to beaker-hostgenerator, most of this code could easily be moved into PSH.

DavidS commented:
To be universally useful, beaker-hostgenerator would also have to have vagrant support.

ferventcoder commented:
> To be universally useful, beaker-hostgenerator would also have to have vagrant support.

Huge :+1:

justinstoller commented:
I don't understand the output of the Travis, it seems the repo isn't set up for it, but Travis thinks it is....

DavidS commented:
Apologies, I went through earlier this week and had IT enable travis on a
bunch of repos where it was missing, but wasn't able to follow up with it.
Please ignore this for now.

On 29 January 2016 at 19:31, Justin Stoller <notifications@github.com>
wrote:

> I don't understand the output of the Travis, it seems the repo isn't set
> up for it, but Travis thinks it is....
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/puppetlabs/modulesync_configs/pull/42#issuecomment-176929060>
> .
>

justinstoller commented:
I would be interested in knowing if there is any other feedback on this?

I believe I've addressed all the comments save for moving this PR to `puppetlabs_spec_helper` which I believe would require adding vagrant support to b-hg, which I believe is well out of the scope of this work (deprecating our internal qe-helpers repo)

justinstoller commented:
ping @DavidS
